src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs

```diff
@@ -19,7 +17,9 @@ public partial class JsonTypeInfo<T>
         internal T? Deserialize(ref Utf8JsonReader reader, ref ReadStack state)
         {
             Debug.Assert(IsConfigured);
-            return EffectiveConverter.ReadCore(ref reader, Options, ref state);
+            bool success = EffectiveConverter.ReadCore(ref reader, out T? result, Options, ref state);
+            Debug.Assert(success, "Should only return false for async deserialization");
```

Comment:
Why is that async specific?

Reply:
`false` is equivalent to `NeedMoreData` which can only happen in the context of async deserialization.